Add the AccessiblePrivateMethods trait

Konamiman opened the pull request:
### All Submissions:

-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

In the new WooCommerce code added in the `src` directory we often use private methods as callbacks for hooks, for readability (to make clear that the method isn't intended to be used from outside the class, other than the hook itself) and maintainability (since making a method public means that we need to keep it forever for backwards compatibility):

```php
add_filter(
    'some_filter',
    function( $argument ) {
        return $this->handle_some_filter( $argument );
    }
);
```

However the problem with this approach is that there isn't an easy way to remove that handler and from time to time people complain about this ([example](https://github.com/woocommerce/woocommerce/pull/30311)).

This pull request introduces an `AccessiblePrivateMethods` trait that allows classes to make private and protected methods accessible from outside of the class, thanks to the magic of `__call` and `__callStatic`. A couple of core `mark_method_as_accessible` and `mark_static_method_as_accessible` methods are provided, but more interstingly, two new `add_action` and `add_filter` methods can be used as direct replacements for the standard functions of the same names, and do the work of both marking the method as accessible from outside and setting up the hook:

```php
class Foobar {
  use AccessiblePrivateMethods;

  public function __construct() {
     self::add_action( 'some_action', [ $this, 'handle_some_action' ] );
  }

  public static function init() {
     self::add_filter('some_filter', [__CLASS__, 'handle_some_filter']);
  }

  private function handle_some_action() {
  }

  private function handle_some_filter() {
  }
}
```

If the supplied callback is not an array, or if the first array element is neither `$this`, `__CLASS__` or another instance of the same class, this won't work as expected (the method won't be marked as accessible), but the corresponding WordPress `add_action` and `add_filter` functions will still be called under the hood.

Worth noting that no special procedure is needed to remove hooks set up with these methods, the regular `remove_action` and `remove_filter` functions provided by WordPress can be used as usual, even from outside of the class.

This pull request also changes the existing hookings to use the trait and the new simplified syntax.

Closes #30311.

### How to test the changes in this Pull Request:

Make sure that the unit tests pass and that the functionality for the classes that have been modified isn't affected (for example, settings pages for product attributes lookup table and custom orders tables are still rendered and work as expected).

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

Konamiman committed:
Add the AccessiblePrivateMethods trait

This trait allows to mark private and protected methods as externally
accessible. There are also utility add_action and add_filter methods
that mark the method as accessible and use it as callback in one
single step.

Konamiman committed:
Use the new AccessiblePrivateMethods trait in existing classes

All the classes that were hooking to private methods using an
anonymous function in the 'src' directory are changed to use the new
AccessiblePrivateMethods trait instead, thus they can be unhooked now.

Konamiman committed:
Small optimization

Konamiman committed:
Add changelog file

Konamiman committed:
Modify yet another existing hook to use AccessiblePrivateMethods trait

lkraav commented:
Could this also help https://github.com/woocommerce/woocommerce/issues/26630 :thinking:

Konamiman commented:
> Could this also help https://github.com/woocommerce/woocommerce/issues/26630 🤔

I don't think so. What this PR allows is to hook to private methods, while the problem described in that issue is that the object doing the hooking isn't accessible (but it's already hooking to public methods).

Konamiman commented:
@lkraav for https://github.com/woocommerce/woocommerce/issues/26630 I have created https://github.com/woocommerce/woocommerce/pull/34156

Konamiman committed:
Add support for visibilizing static methods too.

Konamiman committed:
Merge branch 'trunk' into add/accessible-private-methods-trait

Konamiman committed:
Merge branch 'trunk' into add/accessible-private-methods-trait

Konamiman committed:
Reverse Yoda conditions

Konamiman committed:
Merge branch 'trunk' into add/accessible-private-methods-trait

peterfabian commented on plugins/woocommerce/src/Internal/Traits/AccessiblePrivateMethods.php:
```diff
@@ -0,0 +1,210 @@
+ * $this->add_action('some_action', 'handle_some_action');
+ *
+ * Static methods are supported as well, but in this case the trait methods to use are
+ * 'add_static_ation' and 'add_static_filter':
```
```suggestion
 * 'add_static_action' and 'add_static_filter':
```

peterfabian commented on plugins/woocommerce/src/Internal/Traits/AccessiblePrivateMethods.php:
```diff
@@ -0,0 +1,210 @@
+	 *
+	 * @var array
+	 */
+	private static $_accessible_static_private_methods = array();
```
Is there a reason why these are not declared as arrays?

Konamiman replied:
Not really. I still need to get used to the fact that class members can actually be declared with a type in PHP.

peterfabian replied:
I think I spoke too soon. This is not permitted in 7.2, only later.

peterfabian commented on plugins/woocommerce/src/Internal/Traits/AccessiblePrivateMethods.php:
```diff
@@ -0,0 +1,210 @@
+	 * @return void
+	 */
+	protected function mark_method_as_accessible( string $method_name ): void {
+		// Note that a "is_callable" check would be useless here:
```
```suggestion
		// Note that an "is_callable" check would be useless here:
```

peterfabian commented on plugins/woocommerce/src/Internal/Traits/AccessiblePrivateMethods.php:
```diff
@@ -0,0 +1,210 @@
+
+
+	/**
+	 * Undefined/unaccessible instance method call handler.
```
```suggestion
	 * Undefined/inaccessible instance method call handler.
```

peterfabian commented on plugins/woocommerce/src/Internal/Traits/AccessiblePrivateMethods.php:
```diff
@@ -0,0 +1,210 @@
+	}
+
+	/**
+	 * Undefined/unaccessible static method call handler.
```
```suggestion
	 * Undefined/inaccessible static method call handler.
```

peterfabian commented on plugins/woocommerce/src/Internal/ProductAttributesLookup/LookupDataStore.php:
```diff
@@ -715,4 +639,75 @@ public function lookup_table_has_data(): bool {
+	 * @param string $section_id Settings section identifier.
+	 * @return array New settings configuration array.
+	 */
+	private function add_product_attributes_lookup_table_settings( array $settings, string $section_id ): array {
```
👍 This definitely feels nicer than the whole function body in the add_filter call.

peterfabian commented on plugins/woocommerce/src/Internal/Traits/AccessiblePrivateMethods.php:
```diff
@@ -0,0 +1,210 @@
+	 */
+	protected function add_filter( string $hook_name, $callback, int $priority = 10, int $accepted_args = 1 ): void {
+		if ( is_string( $callback ) && method_exists( $this, $callback ) ) {
+			ArrayUtil::push_once( $this->_accessible_private_methods, $callback );
```
Wouldn't it be better to treat `$this->_accessible_private_methods` as a dictionary with `$callback` as the key here? Perhaps it's a premature optimization, but I always see a red flag when we're iterating over an array looking for a value. 
We're also using the same array in the `__call` further down and that can make function calls and filters/actions unnecessarily slow/expensive, because we have to sequentially search the array(s).

Konamiman replied:
You seem to be right: https://gist.github.com/alcaeus/536156663fac96744eba77b3e133e50a

peterfabian commented on plugins/woocommerce/src/Internal/Traits/AccessiblePrivateMethods.php:
```diff
@@ -0,0 +1,210 @@
+			$callback = array( $this, $callback );
+		} elseif ( is_array( $callback ) && count( $callback ) >= 2 && $callback[0] === $this ) {
+			$this->mark_method_as_accessible( $callback[1] );
+		}
```
I see this has a fallback for other cases we're not handling here, which makes me think if we want `$this->add_filter` and `$this->add_action` to be preferred and generally replace WP equivalents or not?

If not, perhaps we should let the dev know if they're calling this unnecessarily/wrong in some way? At least add something to log or via an exception / `_doing_it_wrong`?

Konamiman replied:
There's no harm in doing `$this->add_action` for callbacks outside the class, and doing so could even be beneficial if in the future we want to implement some kind of hooks registry. So not really "doing wrong" but rather "no need to do that" (but I don't think we have a specific warning for that).

peterfabian commented on plugins/woocommerce/src/Internal/Traits/AccessiblePrivateMethods.php:
```diff
@@ -0,0 +1,210 @@
+	 * $callback can be a standard callable, or a string representing the name of a method in this class.
+	 *
+	 * @param string   $hook_name       The name of the action to add the callback to.
+	 * @param callable $callback        The callback to be run when the action is called.
```
```suggestion
	 * @param callable|string $callback        The callback to be run when the action is called.
```

peterfabian commented on plugins/woocommerce/src/Internal/Traits/AccessiblePrivateMethods.php:
```diff
@@ -0,0 +1,210 @@
+	 * $callback can be a standard callable, or a string representing the name of a method in this class.
+	 *
+	 * @param string   $hook_name       The name of the action to add the callback to.
+	 * @param callable $callback        The callback to be run when the action is called.
```
```suggestion
	 * @param callable|string $callback        The callback to be run when the action is called.
```

peterfabian commented on plugins/woocommerce/src/Internal/Traits/AccessiblePrivateMethods.php:
```diff
@@ -0,0 +1,210 @@
+	 * @param string   $hook_name       The name of the action to add the callback to.
+	 * @param callable $callback        The callback to be run when the action is called.
+	 * @param int      $priority        Optional. Used to specify the order in which the functions
+	 *                                  associated with a particular action are executed.
```
```suggestion
	 *                                  associated with a particular filter are executed.
```

peterfabian commented on plugins/woocommerce/src/Internal/Traits/AccessiblePrivateMethods.php:
```diff
@@ -0,0 +1,210 @@
+	 * @param string   $hook_name       The name of the action to add the callback to.
+	 * @param callable $callback        The callback to be run when the action is called.
+	 * @param int      $priority        Optional. Used to specify the order in which the functions
+	 *                                  associated with a particular action are executed.
```
```suggestion
	 *                                  associated with a particular filter are executed.
```

peterfabian commented on plugins/woocommerce/src/Internal/Traits/AccessiblePrivateMethods.php:
```diff
@@ -0,0 +1,210 @@
+	 * @param string $method_name Method name.
+	 * @return void
+	 */
+	protected function mark_method_as_accessible( string $method_name ): void {
```
I know we're trying to make our methods work in a similar way to WordPress's and for that we don't need any warnings or anything, but the fact that this just silently does nothing when the `$method` doesn't exist makes me a bit uneasy.

Konamiman replied:
An easy solution here could be to return a success/failure boolean.

peterfabian commented on plugins/woocommerce/tests/php/src/Internal/Traits/AccessiblePrivateMethodsTest.php:
```diff
@@ -0,0 +1,439 @@
+	 * @testWith ["protected_return_two"]
+	 *           ["private_return_three"]
+	 *           ["static_protected_return_twenty"]
+	 *           ["static_private_return_thirty"]
```
This `testWith` thing is cool! 👍

peterfabian commented on plugins/woocommerce/tests/php/src/Internal/Traits/AccessiblePrivateMethodsTest.php:
```diff
@@ -0,0 +1,439 @@
+	}
+
+	/**
+	 * @testdox Trying to make a public method accessible with mark_method_as_accessible does nothing.
```
```suggestion
	 * @testdox New add_(static_)action and add_(static_)filter methods can be used to register public class methods as hook callbacks.
```

peterfabian replied:
This actually doesn't test if accessibilizing public method does nothing. I think for that we'd have to check that the trait properties won't include the public methods, no?

Konamiman replied:
"Does nothing" is perhaps a bad description, what I mean is that for practical purposes nothing changes, since `__call(Static)` is never called for public methods.

peterfabian commented on plugins/woocommerce/tests/php/src/Internal/Traits/AccessiblePrivateMethodsTest.php:
```diff
@@ -0,0 +1,439 @@
+			public function __construct( bool $use_string_syntax ) {
+				if ( $use_string_syntax ) {
+					$this->add_action( 'action_handled_privately', 'handle_action' );
+					$this->add_action( 'filter_handled_privately', 'handle_filter' );
```
Should this test `add_filter` instead of `add_action`?

Konamiman replied:
This works because in the end both are the same, but yes, for consistency it should.

peterfabian commented on plugins/woocommerce/tests/php/src/Internal/Traits/AccessiblePrivateMethodsTest.php:
```diff
@@ -0,0 +1,439 @@
+		$this->assertEquals( 'Static filter argument: bar', $filter_result );
+
+		remove_filter( 'filter_handled_privately', array( $sut, 'handle_filter' ) );
+		remove_filter( 'static_filter_handled_privately', array( get_class( $sut ), 'handle_static_filter' ) );
```
should we also test `remove_action`, for completeness?

peterfabian reviewed (commented):
A clever solution and it's implemented nicely, thanks! 
Most of my comments are small nitpicks or doc changes, so nothing huge that isn't working. As mentioned elsewhere, I'm not completely certain if this tradeoff is worth it to get the possibility of unhooking actions/filters, but should the team agree on this approach, this looks solid! 

Unit tests finished fine, as expected also from the CI runs, and I did some spot checks on the functionality you changed, e.g. reviews and it seems to work as expected. Great job!

Konamiman committed:
Simplify the public API of AccessiblePrivateMethods

Now just two static 'add_action' and 'add_filter' static methods
are provided.

Also, the ability to specify just the method name has been removed.

Konamiman committed:
Fix small typos in docblocks for AccessiblePrivateMethods

Also improve one of the tests.

Konamiman committed:
'mark as accessible' methods now return a bool, true if method exists.

Konamiman committed:
Store accessible method names as array keys, instead of array values

This improves the performance of the code ('isset' is way faster
than 'in_array').

Konamiman committed:
Fix additional hook

peterfabian commented:
Can you please rebase and reyodify when you get a chance? (not urgent).

Konamiman committed:
Merge branch 'trunk' into add/accessible-private-methods-trait

Konamiman committed:
Restore Yoda conditions :-(

peterfabian commented:
Should this PR also handle:
- calling of the protected method [over here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/Admin/ProductReviews/Reviews.php#L133-L138)
- calling of private methods [over here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/Admin/Orders/PageController.php#L82-L90)

Konamiman committed:
Use the trait in two more places

peterfabian reviewed (approved):
Looks good now, thanks! (the CI failures seem unrelated)

jwilloth commented:
This commit looks to have broken auto-approving reviews when replying to them, by changing a comparison from "$parent->comment_post_ID == $comment_post_ID" to "$parent->comment_post_ID === $comment_post_ID" (== vs ===) in "woocommerce/src/Internal/Admin/ProductReviews/Reviews.php". 'comment_post_ID' in WP_Comment is a string and $comment_post_ID is cast as an int, so a === comparison returns false. 'wp_ajax_replyto_comment' in wp-admin/includes/ajax-actions.php has the correct form of '=='.

Konamiman commented:
Hi @jwilloth, good catch, thank you. Could you please [turn your comment into an issue](https://github.com/woocommerce/woocommerce/issues/new/choose) with reproduction instructions? Then either ping me or assign the issue directly to me so I can keep track of it. Thanks!
